app/models/shipit/commit.rb

```diff
@@ -194,6 +194,13 @@ def identify_pull_request
       self.pull_request_title = message_parser.pull_request_title unless self[:pull_request_title]
     end
 
+    def deploy_requested_at
+      if pull_request
+        return pull_request.merge_requested_at unless pull_request.canceled?
```

Comment:
@byroot Is the presence of `pull_request` on a commit a certain sign that the merge queue was used?

Reply:
My reading of the code is we only create a pull request locally if added to the merge queue, for example `merge_requested_at `is [`NOT NULL`](https://github.com/Shopify/shipit-engine/blob/ce81004ea2f058815519d3fd581f83e363524ab1/test/dummy/db/schema.rb#L143), so always set.

Reply:
> Is the presence of pull_request on a commit a certain sign that the merge queue was used?

It is, however it doesn't necessarily mean it's the merge queue that merged it.

I think the condition should be:

```
if pull_request.try!(:merged?)
  pull_request.merge_requested_at
else
  created_at
end
```

I don't see a case where a `pull_request`  could have a `merge_requested_at == nil`.

As for `merged?` I need to double check if it's 100% guaranteed it was merged by the merge queue.

Reply:
I double checked, I think we're good.

Reply:
> As for merged? I need to double check if it's 100% guaranteed it was merged by the merge queue.

[Not currently](https://github.com/Shopify/shipit-engine/blob/ce81004ea2f058815519d3fd581f83e363524ab1/app/jobs/shipit/merge_pull_requests_job.rb#L12). I wonder if we could always `cancel!` there...

Reply:
It's another corner case tbf - added to the merge queue, then merged manually without first being removed.

Reply:
Damn, nice catch. I knew we were handling this somewhere, but couldn't find it.



> I wonder if we could always cancel! there...

Hum. Yeah, either cancel, or a new dedicated status. `cancel` would make sense as if you merged manually, you canceled your merge request IMO.

Reply:
It highlights a slight ambiguity in the model name (:bikeshed:), the _merge request_ may have been cancelled (or invalidated?), but the pull request _was_ completed. Let's go with cancel for now, unless we find a need for another state.

Reply:
Indeed. `PullRequest` represent both the GitHub `PullRequest`, and a merge request from the user. 

I definitely though about splitting the two during the original implementation, but though it complicated things. I kinda regret now.

Reply:
> I definitely though about splitting the two during the original implementation, but though it complicated things. I kinda regret now.

I think it could be decoupled with some tweaks to the state machine and schema - the model could represent the the pull request, and the state machine the merge queue. I'll look at that separately.

Happy for this to be merged?

Reply:
> Happy for this to be merged?

Well, I still think checking for `merged?` would be better than `!canceled`.

Reply:
Also right now, you don't fallback to `created_at` if the merge request was canceled.

Reply:
I made that change but didnt push it :doh: done now :)